rfcs/0014-abi-errors.md

```diff
@@ -0,0 +1,456 @@
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Despite the drawbacks of moving more special behavior the compiler, it has to
+shoulder this complexity in Sway since we do not have sufficiently powerful
+metaprogramming features such that developers may be able to define their own
+language features of this caliber.
+
+Given the benefits of having standardized errors, these tradeoffs seem
+acceptable, so long as we allow the solution to be extended in the future to
+be more user defined. The `AbiError` trait is the main vector of such future
+changes in this case.
+
+Letting the errors be populated directly in the bytecode would be a lot simpler,
+but it is not desirable because it extends the size of the binary for no gain.
+
+Moreover, the question of the location of error strings can be asked. Why
+put them in the ABI specification file? ABI files are soon to benefit from
+a registry that will allow users to easily fetch them, they already contain
+type metadata that can be extended, and they ultimately are our equivalent to
+headers, encompassing all the required information to call a contract and get
+a meaningful anwser. Errors are part of making this answer meaningful and it is
+typical for error messages to live in headers.
+
+If we do not specify some kind of error standard, debugging any sort of complex
+Sway error will remain extremely difficult and supporting Sway contracts may be
+a nigh impossible task.
+
+# Prior art
+
+[prior-art]: #prior-art
+
+This RFC is somewhat inspired by Rust's error library ecosystem. We do not
+have enough metaprogramming facilities to allow users to define their own way
+of producing errors, but one can identify two kinds of Rust error handling
+libraries. Those that like `anyhow` care most about letting the user add some
+string based error context. And those like `thiserror` that allow one to build
+complex error types.
+
+We attempt here to cater to both crowds through `panic`'s dual modes.
+
+Moreover, Ethereum style embedding of errors in a `b256` isn't practical for us
+or generally desirable since the FuelVM uses 64 bit error codes and we aim to
+minimize the footprint of errors on deployed contracts.
+
+# Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+It's unclear how much of the existing special error codes produced by the
+standard library (such as `FAILED_ASSERT_SIGNAL`) should be migrated to use this
+mechanism or should remain special behavior.
+
+The extent of the SDK integration of this feature is also quite open ended. But
+implementation of this should give a good enough base to decide how much support
+we want and to extend it in the future.
+
+# Future possibilities
+
+[future-possibilities]: #future-possibilities
+
+In the future we may want to include more metadata for panic invocations or
+error types.
+
+We may also allow error strings to be format strings so that you may use your
+enum's values in the error message.
```

Comment:
Is there a risk that if we don't do format strings from the beginning then adding them later won't be backwards compatible?

Reply:
I can see a purely additive upgrade path from this to format strings: you just record the format strings as normal and change the interpretation from the SDK. We could also add a boolean field right next to them to say they are format strings.

There is a small risk that strings that contain format specifier (i.e.:`{}`) would be affected, but with a new field to mark them this can be done without breakage I think.

Reply:
Assuming we support it later on, means we will support a syntax similar to `thiserror`: `{}` etc. Theoretically, one might use such characters before as a plain text and cause a breaking change afterwards when they get a special meaning.

One way would be to reserve certain pattern in error strings for future improvements but I see that as too restrictive.

I would rather handle this via breaking changes. We are building a tool right now, `forc migrate` that will support in migrations. One day we introduce format strings we warn devs of a potential breaking change in their existing error strings.